API: KMS client interface

ggershinsky opened the pull request.

jackye1995 commented on core/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,115 @@
+   *
+   * @return true if KMS server supports key generation and KmsClient implementation
+   * is interested to leverage this capability. Otherwise, return false - Iceberg will
+   * then generate secret keys locally (using the SecureRandom mechanism) and call
```
nit: {@link java.security.SecureRandom}

jackye1995 commented on core/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,115 @@
+  ByteBuffer unwrapKey(String wrappedKey, String wrappingKeyId);
+
+  /**
+   * Returns the algorithm used by the provider when generating keys, such as AES_256_GCM
```
nit: {@link KeyGenerationAlgorithm#AES_256_GCM}

jackye1995 commented on core/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,115 @@
+  KeyGenerationAlgorithm keyGenerationAlgorithm();
+
+  /**
+   * Returns the spec of the key, such as AES_128
```
nit: {@link KeySpec#AES_128}

jackye1995 commented on core/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,115 @@
+   *
+   * @param properties kms client properties
+   */
+  void setProperties(Map<String, String> properties);
```
Is this the method that initializes the client? In that case we call it `initialize(Map<String, String> properties);` like https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/io/FileIO.java#L64-L69

ggershinsky replied:
yep; actually, it used to be named `initialize` in the MVP code, but then I changed it to `setProperties` in this PR split to make it more flexible and allow for runtime update of credentials (like tokens with expiration). Now I see though that it makes it more complex and less intuitive. And there are many other ways to update the credentials, with or without explicit API. So changing this back to 'initialize`.

jackye1995 commented on core/src/test/java/org/apache/iceberg/encryption/DemoKmsClient.java:
```diff
@@ -0,0 +1,146 @@
+import java.util.Map;
+import javax.crypto.SecretKey;
+
+public class DemoKmsClient implements KmsClient {
```
can we call it something more meaningful like `LocalKeyFileKmsClient`?

ggershinsky replied:
agreed. The main point of this sample is leveraging the standard keystore API, so lets call it KeystoreKmsClient.

jackye1995 commented on core/src/test/java/org/apache/iceberg/encryption/DemoKmsClient.java:
```diff
@@ -0,0 +1,146 @@
+
+public class DemoKmsClient implements KmsClient {
+
+  public static final String KEY_FILE_PATH_PROP = "demo.kms.client.keyfile.path";
```
are these table properties / catalog properties? There are corresponding classes for them at `TableProperties` and `CatalogProperties`.

ggershinsky replied:
these are "kms specific" parameters, consumed by the custom KmsClient code. This parameter is relatively permanent and is not a secret, so can be set as a table property. But the other one (the password) can't. In this demo, it is passed as a Hadoop parameter, which can a fully in-memory transport; alternatively, it can be also passed via an env variable, etc.

jackye1995 reviewed (commented):
I think it's beneficial to also include code about how this class will be used in `EncryptionManager`. Last time during the discussion I remember Ryan was for having these key generation logic directly in the manager itself as a `BaseKmsEncryptionManager`. I think adding that part of the logic here can facilitate the discussion.

I am personally more inclined right now with having a separated KmsClient interface, mostly because this is the service integration point with different KMS services on the market, they don't need to know any logic related to actual Iceberg encryption, this separated interface seems cleaner to me. I would expect only very advanced users to provide an entire custom `EncryptionManager` implementation. @rdblue do you have any comments around the approach to go here?

ggershinsky commented:
> I am personally more inclined right now with having a separated KmsClient interface, mostly because this is the service integration point with different KMS services on the market, they don't need to know any logic related to actual Iceberg encryption, this separated interface seems cleaner to me. I would expect only very advanced users to provide an entire custom `EncryptionManager` implementation. 

I also think this is a simpler and cleaner approach for organizations and their users. One developer/ team would implement the KmsClient class for their organization, integrating with the local KMS and IAM services (the KmsClient interface should be able to support any). Once the implemented class is deployed, any authorized user in the org can write encrypted tables, passing only one parameter - a string with the table key name (later, we'll add column keys). The readers don't need to pass anything. Both writers and readers use auth credentials (KMS-specific). 
Last but not least, this approach is based on the standard "envelope encryption" practice.

> during the discussion I remember Ryan was for having these key generation logic directly in the manager itself as a `BaseKmsEncryptionManager`. I think adding that part of the logic here can facilitate the discussion.

One area where we might need to expand the basic approach, described above, is custom encrypted file streaming. By default, Iceberg will use its built-in encryption for the tables (PME/ORC for data; GcmStream for metadata, and potentially for Avro data; we can also add a built-in AwsS3ClientStream in Iceberg). But if an organization prefers to use its custom encryptors instead of those built in Iceberg, we can load them. I'm sending a new commit to this PR, that adds an interface currently called BaseKmsEncryptionManager. The name and content are a subject for discussion.

> I think it's beneficial to also include code about how this class will be used in `EncryptionManager`. 

We have an internal implementation of EncryptionManager, a class called EnvelopeEncryptionManager that uses a KmsClient plug-in, and leverages the "envelope encryption" practice.

ggershinsky commented on core/src/main/java/org/apache/iceberg/encryption/BaseKmsEncryptionManager.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Module for custom encrypting and decrypting streams for table files.
+ */
+public interface BaseKmsEncryptionManager {
```
It doesn't extend the original EncryptionManager, because this would allow for non-standard key metadata management; this should be done instead via envelope encryption and the KmsClient.

ggershinsky commented:
> @rdblue do you have any comments around the approach to go here?

Ryan, your feedback will be appreciated. This doc section expands on this subject - 
https://docs.google.com/document/d/19O_qiQumz_66CdWLpw38GFJEsUpnNxXckP9rnYIQnCo/edit#heading=h.bq3t7sd49j30

Or, if I try to compress this subject in a few sentences, it would be something like this:
- For new usecases, we create a minimal interface, based on the standard "envelope encryption" practice, and designed to enable connection to KMS/IAM of any org. 
- For existing usecases (if any) and for orgs that for some reason don't want to use envelope encryption (I currently can't see why), we will continue to support the EncryptionManager interface, and will add a missing code that can actually load these custom implementations; but this support shouldn't affect the simplicity/cleanliness of the new interface.
- Finally (a smaller point), the new interface will allow users to plug custom file encryptors into the envelope encryption system, if they are not satisfied with the Iceberg built-in file encryptors (PME, ORC, GcmStream, AwsS3ClientStream).

cc @rdblue

ggershinsky commented:
Ok, I'll proceed with the plan discussed at the slack channel - completing the code updates across the encryption PRs this quarter, so the community can review them (including the interface) next quarter.
cc @rdblue

flyrain commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.Map;
+
+/**
+ * a minimum client interface to connect to a key management service (KMS).
```
Nit: a -> A

flyrain commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+  /**
+   * Wrap a secret key, using a wrapping/master key which is stored in KMS and referenced by an ID.
+   * Wrapping means encryption of the secret key with the master key, and adding optional metadata
+   * that allows KMS to decrypt the secret key in an unwrap call.
```
Nit: unwrap -> unwrapping?

flyrain commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+
+  /**
+   * Wrap a secret key, using a wrapping/master key which is stored in KMS and referenced by an ID.
+   * Wrapping means encryption of the secret key with the master key, and adding optional metadata
```
Question: Does "optional metadata" here refer to the `wrappingKeyId`?

ggershinsky replied:
Correct, the "optional metadata" is not the `wrappingKeyId` . A KMS gets a request to wrap a "secret" using the `wrappingKeyId`. The returned String (a json, typically) contains everything this KMS will need to unwrap the "secret". Always KMS-specific; can be things like the version of the wrapping key, an IV, etc.

flyrain commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+   * @return true if KMS server supports key generation and KmsClient implementation
+   * is interested to leverage this capability. Otherwise, return false - Iceberg will
+   * then generate secret keys locally (using the SecureRandom mechanism) and call
+   * wrapKey to wrap them in KMS.
```
Nit: wrapKey -> {@link #wrapKey(ByteBuffer, String)}

flyrain commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+  ByteBuffer unwrapKey(String wrappedKey, String wrappingKeyId);
+
+  /**
+   * Set kms client properties
```
Nit: -> initialize the KMS client with given properties

flyrain commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+   */
+  void initialize(Map<String, String> properties);
+
+  class KeyGenerationResult {
```
We need a Java doc for this class as well.

flyrain reviewed (commented):
Can we have a unit test for KeyStoreKmsClient? It can demo how the client is used.

andersonm-ibm commented:
> > Can we have a unit test for KeyStoreKmsClient? It can demo how the client is used.
> 
> We will have such a demo / unitest in a later PR, where a table is encrypted using a kms client. However, this rudimentary client won't be a good example for developers. I think the best KmsClient example would be based on Hashicorp Vault (a real KMS; open-source), similar to what we have in Apache Parquet - developed by @andersonm . Maya - will appreciate if you'd be ok with building/contributing such client to Iceberg too :), in a separate pull request.

Yes, @ggershinsky , I am going to work on such an example in a separate PR.

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+   * Wrapping means encryption of the secret key with the master key, and adding optional metadata
+   * that allows KMS to decrypt the secret key in an unwrap call.
+   *
+   * @param key            a secret key being wrapped
```
Nit; not vertical alignment

RussellSpitzer commented on core/src/test/java/org/apache/iceberg/encryption/kms/KeyStoreKmsClient.java:
```diff
@@ -0,0 +1,113 @@
+  @Override
+  public void initialize(Map<String, String> properties) {
+    String keyFilePath = properties.get(KEY_FILE_PATH_PROP);
+    if (null == keyFilePath) {
```
We use `Precondition.checkArg` for things like this

RussellSpitzer replied:
```Preconditions.checkNotNull(properties.get(KEY_FILE_PATH_PROP),  PROP + " cannot be null)```

RussellSpitzer commented on core/src/test/java/org/apache/iceberg/encryption/kms/MemoryMockKMS.java:
```diff
@@ -0,0 +1,68 @@
+
+  @Override
+  public String wrapKey(ByteBuffer key, String wrappingKeyId) {
+    // keytool keeps key names in lower case
```
comment is here but we don't actually make the key lower case here

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+   * then generate secret keys locally (using the SecureRandom mechanism) and call
+   * wrapKey to wrap them in KMS.
+   */
+  boolean supportsKeyGeneration();
```
Should we do a default implementation of false here?

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,96 @@
+   * @param wrappingKeyId   a key ID that represents a wrapping key stored in KMS
+   * @return                a pair of plaintext and wrapped key encrypted with the given wrappingKeyId
+   */
+  KeyGenerationResult generateKey(String wrappingKeyId);
```
This one maybe should have a default to "unsupported operation exception"?

ggershinsky commented:
> The only real concern I have is that the test class implementations are here but we don't make sure they work. That's probably fine since the actual usage of this interface will exercise those apis? If that's coming up then +1 from me

Yep, these implementations are placed in the test module; they'll serve as an util for a unitest that encrypts/decrypts a table using these test kms classes. This unitest is coming up in the 5th PR in this series, where we connect all parts to enable table encryption.
Though, I can add a unitest just for these tester kms classes, should be easy - but I wonder if there is a precedent / need in Iceberg to test classes that reside in a test module (not in a main module), and are built for support other unitests.

ggershinsky commented:
the commit addresses the comments in the last round. I also made an extra change in the Keystore client (using env var instead of hadoop prop for the password; this is safer and more intuitive).

RussellSpitzer commented:
@jackye1995 Are you +1 on this? I think we have consensus to start getting these PR's in

@ggershinsky looks like the tests are currently failing, could you rebase so we can be good to go?

jackye1995 reviewed (approved):
Looks good to me!

ggershinsky commented:
> looks like the tests are currently failing

Hi @RussellSpitzer , this PR depends on https://github.com/apache/iceberg/pull/2638

ggershinsky commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,103 @@
+   * @param wrappingKeyId a key ID that represents a wrapping key stored in KMS
+   * @return wrapped key material
+   */
+  String wrapKey(ByteBuffer key, String wrappingKeyId);
```
KMS systems typically return a JSON String with a wrapped key. If for some reason a KMS returns a raw byte buffer, it should be converted to a String via base64.

RussellSpitzer commented on core/src/test/java/org/apache/iceberg/encryption/kms/KeyStoreKmsClient.java:
```diff
@@ -0,0 +1,118 @@
+    Preconditions.checkNotNull(keystorePath, KEYSTORE_FILE_PATH_PROP + " must be set in hadoop or table " +
+        "properties");
+
+    String keystorePassword = System.getenv(KEYSTORE_PASSWORD_ENV_VAR);
```
I know this is just a demo class, but won't this intern the keystorePassword in memory? Not sure if that's a concern or anything I am still an encryption novice.

ggershinsky replied:
hey @RussellSpitzer , thanks for merging this . To answer the question - this is not a concern in a basic threat model, where we don't trust the storage domain, but we do trust the domain of compute nodes - we have to, because the node memory sees the sensitive data, the encryption keys, etc. The meaning of "trust" is a deep hole.., but to give a simple example, the compute nodes run in-house, while the storage backend is on another company's platform.

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/encryption/KmsClient.java:
```diff
@@ -0,0 +1,103 @@
+  /**
+   * Some KMS systems support generation of secret keys inside the KMS server.
+   *
+   * @return true if KMS server supports key generation and KmsClient implementation
```
nit: I don't think the "and KmsClient Implementation .... " part makes sense to me here? Is the statement a "Return true IFF the Kms Server can generate keys and this client should use that capability?"

RussellSpitzer reviewed (commented):
My only question on this is whether or not we should have tests for this interface or do we not yet have the components required to do a meaningful test yet?

RussellSpitzer commented:
Oh sorry I forgot we talked about this before. Everyone is on board so let's move forward.

RussellSpitzer commented:
Thanks @ggershinsky for the pr and 
@jackye1995 + @flyrain and @shangxinli  for taking a reviewing!
